src/Symfony/Component/Utf8/Utf8Trait.php

```diff
@@ -0,0 +1,363 @@
+                $string .= chr(0xF0 | $code >> 18).chr(0x80 | $code >> 12 & 0x3F).chr(0x80 | $code >> 6 & 0x3F).chr(0x80 | $code & 0x3F);
+            }
+        }
+
+        $instance = new static();
+        $instance->string = $string;
+
+        return $instance;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function __toString()
+    {
+        return $this->string;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function isEmpty(): bool
+    {
+        return '' === $this->string;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function explode(string $delimiter, int $limit = null): array
+    {
+        if ('' === $delimiter) {
+            throw new InvalidArgumentException('Passing an empty delimiter is not supported by this method. Use getIterator() method instead.');
+        }
+
+        $chunks = array();
+        foreach (preg_split('/'.preg_quote($delimiter).'/u', $this->string, $limit ?: -1) as $i => $string) {
+            $chunks[$i] = $chunk = clone $this;
+            $chunk->string = $string;
+        }
+
+        return $chunks;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function toLowerCase(): self
+    {
+        $result = clone $this;
+        $result->string = mb_strtolower($this->string, 'UTF-8');
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function toUpperCase(): self
+    {
+        $result = clone $this;
+        $result->string = mb_strtoupper($this->string, 'UTF-8');
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function toUpperCaseFirst($allWords = false): self
+    {
+        $callback = function ($matches) {
+            return mb_convert_case($matches[1], MB_CASE_TITLE, 'UTF-8');
+        };
+
+        if ($allWords) {
+            $string = preg_replace_callback('/\b(.)/u', $callback, $this->string);
+        } else {
+            $capitalLetter = mb_substr($this->string, 0, 1, 'UTF-8');
+            $string = preg_replace_callback('/\b(.)/u', $callback, $capitalLetter).substr($this->string, strlen($capitalLetter));
+        }
+
+        $result = clone $this;
+        $result->string = $string;
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function toFoldedCase(bool $full = true): self
+    {
+        $string = str_replace(self::$commonCaseFold[0], self::$commonCaseFold[1], $this->string);
+
+        if ($full) {
+            static $fullCaseFold = false;
+            $fullCaseFold or $fullCaseFold = require __DIR__.'/Resources/data/caseFolding_full.php';
+            $string = str_replace($fullCaseFold[0], $fullCaseFold[1], $string);
+        }
+
+        $result = clone $this;
+        $result->string = mb_strtolower($string, 'UTF-8');
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function trim(string $charsList = null): self
+    {
+        if (!preg_match('//u', $charsList)) {
```

Comment:
You said it! The reason why this logic is duplicated is to avoid extra method calls. We want this set of classes to be as efficient as possible. That also explains why all classes are final so that we can change them later to improve them and ensure our BC promise.